src/workerd/io/actor-sqlite.c++

```diff
@@ -177,6 +235,29 @@ void ActorSqlite::requireNotBroken() {
   }
 }
 
+void ActorSqlite::maybeDeleteDeferredAlarm() {
+  KJ_IF_SOME(d, currentAlarmTime.tryGet<DeferredAlarmDelete>()) {
+    // TODO(now): Do we need to do anything special, if we've already started flushing the alarm?
+    d.status = DeferredAlarmDelete::Status::READY;
+    // TODO(now): Do we need to do anything special to ensure that the commit code for this sqlite
```

Comment:
@kentonv is more familiar with this than me, but it sure looks to me like we'd be better off going through an `ActorSqlite` code path that will ensure this gets flushed properly in accordance with output gates. I would have expected this to go through `ActorSqlite::setAlarm` and/or the code path in `commitImpl`, since as is this looks like it won't do anything more than update the sqlite DB without ever calling out to the `ActorSqliteHooks`.

Reply:
My understanding is that calling `metadata.setAlarm(kj::none);` will run the sqlite query to update the alarm state in the sqlite database, and that this will cause sqlite to call `ActorSqlite::onWrite()`, which should set up an implicit transaction to be committed on a later run of the event loop, which would then invoke the necessary hooks functions?

But looking at the current implementation of `onWrite()`, it looks like it would only set up an implicit transaction if there is not an ongoing explicit transaction... which I guess means that someone might be able to start an explicit transaction, and any alarm handler invocations that finish during the transaction would not delete their alarms until the explicit transaction was committed.

It seems kind of weird for an object to be holding open an explicit transaction across alarm handler boundaries, but I'm not sure we disallow it, so maybe this case needs to be handled, too?

Reply:
Oh, nice, thanks for calling out how that `onWrite()` callback is hooked up.

I'm not 100% certain on this without digging back in, but it looks like an `ExplicitTxn` is only ever created by the `startTransaction()` function, which is only called from `DurableObjectStorage::transaction`, which wraps the `startTransaction()` call in a `blockConcurrencyWhile()`, which should eliminate any concerns about an alarm handler finishing up while some other request is doing a txn. So I think we're ok here.

And I was going to say I was worried about raw `BEGIN` statements being sent by the user, but then saw that we have the `Regulator` class configured to disallow those from JS. So I think we're ok here.

And on the plus side, if something goes wrong and we end up not deleting an alarm time, we'll just harmlessly re-run the alarm, so it's not the highest of stakes.

Reply:
Hmm... that's my intuition of how sqlite uses `onWrite()`, anyway, but thinking more about it, I should probably verify the behavior.  One thing I'm not quite sure of is how it interacts with transactions... I assume that:

- if someone calls `setAlarm()` within a transaction, they don't expect the alarm handler to actually be called until after the transaction is committed (and any transactions enclosing it).  I'm pretty sure the current PR does this, but should probably add a test for it.
- if someone calls `setAlarm()` within a transaction, then rolls back the transaction, they expect the alarm state to revert to the pre-transaction value.  I think the sqlite alarm state will revert correctly in the current PR, but not sure the cached value in `currentAlarmTime` would.  Which I guess is an argument for dropping the caching and just querying the sqlite value directly.  (and of course would also benefit from a test case)

Are these assumptions correct?

> which wraps the startTransaction() call in a blockConcurrencyWhile(), which should eliminate any concerns about an alarm handler finishing up while some other request is doing a txn. So I think we're ok here.

Ah, cool... I was thinking that maybe `blockConcurrencyWhile()` only prevented the start of new JS events, but still allowed the completion of ongoing async C++ promises, and wasn't sure if maybe that would allow things like the alarm handler cleanup to sneak in.

Reply:
> if someone calls setAlarm() within a transaction, they don't expect the alarm handler to actually be called until after the transaction is committed (and any transactions enclosing it). I'm pretty sure the current PR does this, but should probably add a test for it.

Agreed.

> if someone calls setAlarm() within a transaction, then rolls back the transaction, they expect the alarm state to revert to the pre-transaction value. I think the sqlite alarm state will revert correctly in the current PR, but not sure the cached value in currentAlarmTime would. Which I guess is an argument for dropping the caching and just querying the sqlite value directly. (and of course would also benefit from a test case)

Agreed, and good catch!

> Ah, cool... I was thinking that maybe blockConcurrencyWhile() only prevented the start of new JS events, but still allowed the completion of ongoing async C++ promises, and wasn't sure if maybe that would allow things like the alarm handler cleanup to sneak in.

`blockConcurrencyWhile()` prevents any events from being delivered to JS, including both new incoming requests but also responses to previously issued outgoing I/O. We don't necessarily stop ongoing async C++ work from *happening*, but we do prevent it from waking JS code back up.